server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/VariableWidthHistogramAggregator.java

```diff
@@ -293,17 +283,33 @@ public CollectionPhase collectValue(LeafBucketCollector sub, int doc, double val
                 collectBucket(sub, doc, numClusters - 1);
 
                 if(val > clusterCentroids.get(bucketOrd)){
-                    // Insert just ahead of bucketOrd so that the array remains sorted
+                    /*
+                     * If the new value is bigger than the nearest bucket then insert
+                     * just ahead of bucketOrd so that the array remains sorted.
+                     */
                     bucketOrd += 1;
                 }
                 moveLastCluster(bucketOrd);
+                updateAvgBucketDistanceIfModified(bucketOrd);
             } else {
                 addToCluster(bucketOrd, val);
                 collectExistingBucket(sub, doc, bucketOrd);
+                updateAvgBucketDistanceIfModified(bucketOrd);
             }
             return this;
         }
 
+        private void updateAvgBucketDistanceIfModified(int modifiedBucketOrd) {
+            if (modifiedBucketOrd == 0 || modifiedBucketOrd == numClusters - 1) {
+                updateAvgBucketDistance();
+            }
+        }
+
+        private void updateAvgBucketDistance() {
+            // Centroids are sorted so the average distance is the difference between the first and last.
```

Comment:
This is a really interesting observation! It took me a bit of thinking to convince myself that this is equivalent to the previous equation, but I agree that this makes sense.

The next question I had was whether this is the right metric to use after all, since this means that if a bucket is placed between the first and last bucket, its actual location doesn't affect the `avgBucketDistance`.

But upon further consideration I think it definitely makes sense, since this is still just measuring the distance between buckets.